pkg/remotewrite/trend.go

```diff
@@ -1,84 +1,109 @@
+	// store a counter for the number of reported values and gauges to keep
+	// track of aggregated values. Also store a sum of the values to allow
+	// the calculation of moving averages.
+	// TODO: when Prometheus implements support for sparse histograms, re-visit this implementation
 
-func (t *trendSink) Add(s metrics.Sample) {
-	// insert into sorted array instead of sorting anew on each addition
-	index := sort.Search(len(t.Values), func(i int) bool {
-		return t.Values[i] > s.Value
-	})
-	t.Values = append(t.Values, 0)
-	copy(t.Values[index+1:], t.Values[index:])
-	t.Values[index] = s.Value
+	tg := &trendAsGauges{
+		series:    make([]*prompb.TimeSeries, 0, 8),
+		labels:    MapSeries(series),
+		timestamp: t.UnixMilli(),
+	}
+	tg.CacheNameIndex()
+
+	tg.Append("count", float64(sink.Count))
+	tg.Append("sum", sink.Sum)
+	tg.Append("min", sink.Min)
+	tg.Append("max", sink.Max)
+	tg.Append("avg", sink.Avg)
+	tg.Append("med", sink.P(0.5))
+	tg.Append("p90", sink.P(0.9))
+	tg.Append("p95", sink.P(0.95))
+	return tg.series
+}
+
+type trendAsGauges struct {
+	// series is the slice of the converted TimeSeries.
+	series []*prompb.TimeSeries
 
-	t.Count += 1
-	t.Sum += s.Value
-	t.Avg = t.Sum / float64(t.Count)
+	// labels are the shared labels between all the Gauges.
+	labels []*prompb.Label
 
-	if s.Value > t.Max {
-		t.Max = s.Value
+	// timestamp is the shared timestamp in ms between all the Gauges.
+	timestamp int64
+
+	// ixname is the slice's index
+	// of the __name__ Label item.
+	//
+	// 16 bytes should be enough for the max length
+	// an higher value will probably generate
+	// serious issues in other places.
+	ixname uint16
+}
+
+func (tg *trendAsGauges) Append(suffix string, v float64) {
+	ts := &prompb.TimeSeries{
+		Labels:  make([]*prompb.Label, len(tg.labels)),
+		Samples: make([]*prompb.Sample, 1),
 	}
-	if s.Value < t.Min || t.Count == 1 {
-		t.Min = s.Value
+	for i := 0; i < len(tg.labels); i++ {
+		ts.Labels[i] = &prompb.Label{
+			Name:  tg.labels[i].Name,
+			Value: tg.labels[i].Value,
+		}
 	}
+	ts.Labels[tg.ixname].Value += "_" + suffix
 
-	if (t.Count & 0x01) == 0 {
-		t.Med = (t.Values[(t.Count/2)-1] + t.Values[(t.Count/2)]) / 2
-	} else {
-		t.Med = t.Values[t.Count/2]
+	ts.Samples[0] = &prompb.Sample{
+		Timestamp: tg.timestamp,
+		Value:     v,
 	}
+	tg.series = append(tg.series, ts)
 }
 
-func (t *trendSink) P(pct float64) float64 {
-	switch t.Count {
-	case 0:
-		return 0
-	case 1:
-		return t.Values[0]
-	default:
-		// If percentile falls on a value in Values slice, we return that value.
-		// If percentile does not fall on a value in Values slice, we calculate (linear interpolation)
-		// the value that would fall at percentile, given the values above and below that percentile.
-		i := pct * (float64(t.Count) - 1.0)
-		j := t.Values[int(math.Floor(i))]
-		k := t.Values[int(math.Ceil(i))]
-		f := i - math.Floor(i)
-		return j + (k-j)*f
+// FindNameIndex finds the __name__ label's index
```

Comment:
```suggestion
// CacheNameIndex finds the __name__ label's index
```

Reply:
Apparently, we have something to fix in the linter...